src/tlc_tools/metrics.py

```diff
@@ -0,0 +1,212 @@
+
+    # Calculate uniqueness scores
+    uniqueness_scores = np.zeros(len(embeddings), dtype=float)
+
+    for label in unique_labels:
+        indices = label_to_indices[label]
+        label_embeddings = embeddings[indices]
+
+        # Compute pairwise distances within each class
+        pairwise_distances = squareform(pdist(label_embeddings))
+
+        # Calculate uniqueness as the mean distance of each image to others in its class
+        mean_distances = pairwise_distances.mean(axis=1)
+
+        # Assign the uniqueness score based on mean distances
+        for idx, mean_distance in zip(indices, mean_distances):
+            uniqueness_scores[idx] = mean_distance
+
+    return uniqueness_scores.tolist()
+
+
+def traversal_index(embeddings):
+    embeddings = np.array(embeddings)
+
+    # Normalize embeddings
+    norms = np.linalg.norm(embeddings, axis=1, keepdims=True)
+    normalized_embeddings = embeddings / (norms + 1e-12)
+
+    # Unique embeddings determination
+    unique_embeddings, unique_indices, inverse_indices = np.unique(
+        normalized_embeddings, axis=0, return_index=True, return_inverse=True
+    )
+
+    # Find the center of unique embeddings
+    center_embedding = unique_embeddings.mean(axis=0)
+    distances_from_center = np.linalg.norm(unique_embeddings - center_embedding, axis=1)
+    center_index_in_unique = np.argmin(distances_from_center)
+
+    # Initialize traversal order with the index of the central unique embedding
+    traversal_order = [center_index_in_unique]
+    selected_embeddings = set(traversal_order)
+
+    # Iteratively select the farthest unpicked unique embedding
+    remaining_count = len(unique_embeddings) - 1
+    while remaining_count > 0:
+        last_embedding = unique_embeddings[traversal_order[-1]]
+        distances = np.linalg.norm(unique_embeddings - last_embedding, axis=1)
+
+        for idx in np.argsort(distances)[::-1]:  # Start from farthest
+            if idx not in selected_embeddings:
+                traversal_order.append(idx)
+                selected_embeddings.add(idx)
+                remaining_count -= 1
+                break
+
+    # Add remaining unique indices in random order if not yet selected
+    remaining_unique = [i for i in range(len(unique_embeddings)) if i not in traversal_order]
+    np.random.shuffle(remaining_unique)
+    traversal_order.extend(remaining_unique)
+
+    # Map traversal_order back to original embeddings indices
+    final_traversal_order = [unique_indices[idx] for idx in traversal_order]
+
+    # Add duplicates in random order
+    duplicate_indices = [i for i in range(len(embeddings)) if i not in unique_indices]
+    np.random.shuffle(duplicate_indices)
+    final_traversal_order.extend(duplicate_indices)
+
+    return [int(idx) for idx in final_traversal_order]
+
+IMAGE_METRICS = Literal["width", "height", "brightness", "contrast", "sharpness", "average_red", "average_green", "average_blue"]
```

Comment:
This was not so straight forward, as Literal[ALL_IMAGE_METRICS] will not be valid since it is an expression in the literal.. If I have time I will try to find a way to not repeat myself.